near-lake-framework/src/streamer.rs

```diff
@@ -0,0 +1,230 @@
+use aws_sdk_s3::Client;
+
+use futures::stream::StreamExt;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::SendError;
+
+use near_lake_primitives::near_indexer_primitives;
+
+pub(crate) fn streamer(
+    config: crate::types::Lake,
+) -> (
+    tokio::task::JoinHandle<Result<(), anyhow::Error>>,
+    mpsc::Receiver<near_indexer_primitives::StreamerMessage>,
+) {
+    let (sender, receiver) = mpsc::channel(config.blocks_preload_pool_size);
+    (tokio::spawn(start(sender, config)), receiver)
+}
+
+fn stream_block_heights<'a: 'b, 'b>(
+    s3_client: &'a Client,
+    s3_bucket_name: &'a str,
+    mut start_from_block_height: crate::types::BlockHeight,
+    number_of_blocks_requested: usize,
+) -> impl futures::Stream<Item = u64> + 'b {
+    async_stream::stream! {
+        loop {
+            tracing::debug!(target: crate::LAKE_FRAMEWORK, "Fetching a list of blocks from S3...");
+            match crate::s3_fetchers::list_blocks(
+                s3_client,
+                s3_bucket_name,
+                start_from_block_height,
+                number_of_blocks_requested
+            )
+            .await {
+                Ok(block_heights) => {
+                    if block_heights.is_empty() {
+                        tracing::debug!(
+                            target: crate::LAKE_FRAMEWORK,
+                            "There are no newer block heights than {} in bucket {}. Fetching again in 2s...",
+                            start_from_block_height,
+                            s3_bucket_name,
+                        );
+                        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
+                        continue;
+                    }
+                    tracing::debug!(
+                        target: crate::LAKE_FRAMEWORK,
+                        "Received {} newer block heights",
+                        block_heights.len()
+                    );
+
+                    start_from_block_height = *block_heights.last().unwrap() + 1;
+                    for block_height in block_heights {
+                        tracing::debug!(target: crate::LAKE_FRAMEWORK, "Yielding {} block height...", block_height);
+                        yield block_height;
+                    }
+                }
+                Err(err) => {
+                    tracing::warn!(
+                        target: crate::LAKE_FRAMEWORK,
+                        "Failed to get block heights from bucket {}: {}. Retrying in 1s...",
+                        s3_bucket_name,
+                        err,
+                    );
+                    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                }
+            }
+        }
+    }
+}
+
+async fn fast_fetch_block_heights(
+    pending_block_heights: &mut std::pin::Pin<&mut impl tokio_stream::Stream<Item = u64>>,
+    limit: usize,
+    await_for_at_least_one: bool,
```

Comment:
`await_for_at_least_one_block`?

Reply:
Good question, I will check it out and provide doc-strings